lib/Infer/DataflowPruning.cpp

```diff
@@ -0,0 +1,330 @@
+#include "souper/Infer/DataflowPruning.h"
+
+namespace souper {
+namespace {
+using souper::dataflow::EvalValue;
+using souper::dataflow::ValueCache;
+
+EvalValue EvaluateInst(souper::Inst* Inst, std::vector<EvalValue> args) {
+  switch (Inst->K) {
+    case souper::Inst::Const:
+      return {Inst->Val, false, false};
+    case souper::Inst::UntypedConst:
+      return {Inst->Val, false, false};
+    case souper::Inst::Var:
+      llvm_unreachable("Should get value from cache without reaching here");
+    case souper::Inst::Add:
+      return {args[0].Val + args[1].Val, false, false};
+  //   case AddNSW:
+  //     return "addnsw";
+  //   case AddNUW:
+  //     return "addnuw";
+  //   case AddNW:
+  //     return "addnw";
+    case souper::Inst::Sub:
+      return {args[0].Val - args[1].Val, false, false};
+  //   case SubNSW:
+  //     return "subnsw";
+  //   case SubNUW:
+  //     return "subnuw";
+  //   case SubNW:
+  //     return "subnw";
+    case souper::Inst::Mul:
+      return {args[0].Val * args[1].Val, false, false};
+
+  //   case MulNSW:
+  //     return "mulnsw";
+  //   case MulNUW:
+  //     return "mulnuw";
+  //   case MulNW:
+  //     return "mulnw";
+    case souper::Inst::UDiv:
+      return {args[0].Val.udiv(args[1].Val), false, false};
+    case souper::Inst::SDiv:
+      return {args[0].Val.sdiv(args[1].Val), false, false};
+  //   case UDivExact:
+  //     return "udivexact";
+  //   case SDivExact:
+  //     return "sdivexact";
+    case souper::Inst::URem:
+      return {args[0].Val.urem(args[1].Val), false, false};
+    case souper::Inst::SRem:
+      return {args[0].Val.srem(args[1].Val), false, false};
+    case souper::Inst::And:
+      return {args[0].Val & args[1].Val, false, false};
+    case souper::Inst::Or:
+      return {args[0].Val | args[1].Val, false, false};
+    case souper::Inst::Xor:
+      return {args[0].Val ^ args[1].Val, false, false};
+    case souper::Inst::Shl:
+      return {args[0].Val << args[1].Val, false, false};
+  //   case ShlNSW:
+  //     return "shlnsw";
+  //   case ShlNUW:
+  //     return "shlnuw";
+  //   case ShlNW:
+  //     return "shlnw";
+    case souper::Inst::LShr:
+      return {args[0].Val.lshr(args[1].Val), false, false};
+  //   case LShrExact:
+  //     return "lshrexact";
+    case souper::Inst::AShr:
+      return {args[0].Val.ashr(args[1].Val), false, false};
+      //   case AShrExact:
+  //     return "ashrexact";
+    case souper::Inst::Select: {
+      if (args[0].Unavailable || args[0].Poison) {
+        return args[0];
+      } else {
+        bool cond = args[0].Val.getLimitedValue();
```

Comment:
this isn't a good use of getLimitedValue, use getBoolValue

Reply:
Fixed